grafana/opensearch-datasource
# Service Map: Build Stats request and process responses to dataframes

Split the feature into multiple PRs for ease of reviewing. The feature is not enabled by default but 'hidden' behind a provisional feature flag.
After [getting operationNames and serviceNames from the service map prefetch](https://github.com/grafana/opensearch-datasource/pull/362) request, we build the query for getting the stats (throughput, error_rate, avg_latency) for each node (i.e. service). 

This PR implements:
- Detecting the stats query in query_processing.go, and splitting the query in two - one for service map (to get the nodes and edges) and one for stats (latency etc.). They must be split because their aggregations are made on two different sets of filtered documents, so they cannot be combined into one query
- Building the stats query, with filters and aggregations (including filters for traceId in case one one trace is queried)
- in response_processing.go, combining the two results into dataframes that the Node Graph Grafana visualization can read. 


A few things to note: 
- This builds queries for otel + Open Search data prepper. Open Search can [store traces in raw Jaeger format](https://opensearch.org/docs/latest/observing-your-data/trace/index/#trace-analytics-with-jaeger-data), in which case the aggregations and responses should look slightly different. This is something we intend to implement in the future.
- We send the service map index request an extra time, together with the Stats request. This isn't necessary, but we thought it's the cleanest solution - we would otherwise have to encode the whole response in the Stats query `serviceMapInfo` field, to be later processed together with the response from Stats query. The current backend architecture makes this even more difficult because we cannot intercept raw response from Open Search anywhere. In order to mitigate this in the future, we intend to add client-side caching for the service map prefetch response (along with refactoring the backend querying, but that's a significant chunk of work that might take a while)

**Testing**

1. enable openSearchNodeGraph feature toggle in Grafana config
2. run the Data Prepper/examples/jaeger-hotrod
3. generate some traces in the browser app
4. select query type: Traces and switch on the Service Map button in the query editor
5. Run the query 
6. In explore, there will be a Node Graph panel along with the table panel. In dashboard panel, you'll be able to switch to the node graph panel
7. If you query a particular trace ID from the table `traceId: xyz` and select Service Map switch, you can see the Node Graph for that particular trace along with the span timeline
